crates/sound/src/stream.rs

```diff
@@ -251,33 +252,67 @@ impl std::fmt::Debug for Buffer {
     }
 }
 
-impl Buffer {
-    pub fn new(data_descriptor: virtio_queue::Descriptor, message: Arc<IOMessage>) -> Self {
+impl Request {
+    pub fn new(len: usize, message: Arc<IOMessage>) -> Self {
         Self {
+            len,
             pos: 0,
-            data_descriptor,
             message,
+            pos_curr_desc: 0,
+            // payload begins after the header descriptor
+            idx_curr_desc: 1,
         }
     }
 
-    pub fn consume(&self, buf: &mut [u8]) -> Result<u32> {
-        let addr = self.data_descriptor.addr();
-        let offset = self.pos as u64;
-        let len = self
-            .message
-            .desc_chain
-            .memory()
-            .read(
-                buf,
-                addr.checked_add(offset)
-                    .expect("invalid guest memory address"),
-            )
-            .map_err(|_| Error::DescriptorReadFailed)?;
-        Ok(len as u32)
+    pub fn consume(&mut self, buf: &mut [u8]) -> Result<u32> {
```

Comment:
Looks like everyone has to work around that something like https://github.com/rust-vmm/vm-virtio/pull/33 is missing. It looks like every device falls into two categories. Either it implements its own descriptor chain walking or ends up glossing over the fact that descriptors may be sliced up arbitrarily. For SCSI we already carry descriptor chain walking. This would be another variant.

I think we should just provide it as part of vm-virtio and remove our own attempts of doing it from vhost-device.

Reply:
That'd be nice. I suggest importing the descriptor_utils from crosvm, which is quite powerful and flexible.